src/dhtproto/client/request/internal/GetHashRange.d

```diff
@@ -208,229 +193,194 @@ public struct GetHashRange
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            AddrPort(addr, port), min, max);
-                        break;
-
-                    case ChangeHashRange:
-                        hash_t new_min, new_max;
-                        this.conn.message_parser.parseBody(payload,
-                            new_min, new_max);
-
-                        auto shared_resources = SharedResources.fromObject(
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            this.conn.remote_address, new_min, new_max);
-                        break;
-
-                    default:
-                        log.error("Unknown message code {} received", msg_type);
-                        this.conn.shutdownWithProtocolError("Message parsing error");
-                }
+                Const!(void)[] payload_slice = payload;
+                min = *this.conn.message_parser.getValue!(hash_t)(payload_slice);
+                max = *this.conn.message_parser.getValue!(hash_t)(payload_slice);
             }
         );
+        this.resources.node_hash_ranges.updateNodeHashRange(
+            this.conn.remote_address, min, max);
 
-        return State.Receiving;
+        while ( true )
+        {
+            this.conn.receive(
+                ( in void[] payload )
+                {
+                    Const!(void)[] payload_slice = payload;
+                    auto msg_type = *this.conn.message_parser.
+                        getValue!(MessageType)(payload_slice);
+
+                    with ( MessageType ) switch ( msg_type )
+                    {
+                        case NewNode:
+                            typeof(AddrPort.naddress) addr;
+                            typeof(AddrPort.nport) port;
+                            hash_t min, max;
+                            this.conn.message_parser.parseBody(payload_slice,
+                                addr, port, min, max);
+
+                            this.resources.node_hash_ranges.updateNodeHashRange(
+                                AddrPort(addr, port), min, max);
+                            break;
+
+                        case ChangeHashRange:
+                            hash_t new_min, new_max;
+                            this.conn.message_parser.parseBody(payload_slice,
+                                new_min, new_max);
+
+                            this.resources.node_hash_ranges.updateNodeHashRange(
+                                this.conn.remote_address, new_min, new_max);
+                            break;
+
+                        default:
+                            log.error("Unknown message code {} received", msg_type);
+                            this.conn.shutdownWithProtocolError("Message parsing error");
```

Comment:
Do you need to throw here?

```
throw this.conn.shutdownWithProtocolError(...)
``` ?

Reply:
OHHH! Shit. I didn't know that method _returns_ an exception, rather than actually throwing one :grimacing:

Reply:
See https://github.com/sociomantic-tsunami/swarm/pull/292.

Reply:
Updated.